### PR TITLE
feat(pubsublite): implementation of user-facing and helper PSL types

### PR DESCRIPTION
**Pull request by 18suresha**
This PR implements the required data files for Pub/Sub Lite and helper data representations as seen in other client libraries.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591)

### CODE REVIEW

**Review by dpcollins-google — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591)* status: 0 of 12 files reviewed, 11 unresolved discussions (waiting on @18suresha and @dpcollins-google)

___
*[google/cloud/pubsublite/message_metadata.h, line 33 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MywzmsF6dfSApt1azTE:-MywzmsG1Gf6jDXMDqA9:b-wrtx4n) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/message_metadata.h#L33)):*
> ```objc
>       : partition_{partition}, cursor_{std::move(cursor)} {}
> 
>   std::int64_t GetPartition() const { return partition_; }
> ```

nit. just Partition(), Get isn't required



_Code quote:_
```objc
Get
```
___
*[google/cloud/pubsublite/internal/cloud_region.h, line 30 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Myx-mgFEZg0Dh8bglJh:-Myx-mgFEZg0Dh8bglJi:bj5zusq) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/internal/cloud_region.h#L30)):*
> ```objc
>  * A wrapped string representing a Google Cloud region.
>  */
> class CloudRegion {
> ```

this can just be a struct



_Code quote:_
```objc
CloudRegion
```
___
*[google/cloud/pubsublite/internal/cloud_region.h, line 30 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Myx-o4X0jhUBrz7VA3e:-Myx-o4X0jhUBrz7VA3f:b-4e4u9w) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/internal/cloud_region.h#L30)):*
> ```objc
>  * A wrapped string representing a Google Cloud region.
>  */
> class CloudRegion {
> ```

Add a Parse static member function
___
*[google/cloud/pubsublite/internal/cloud_zone.h, line 33 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Myx-jTA-4j_XwbPqWF0:-Myx-jTA-4j_XwbPqWF1:b-i0v986) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/internal/cloud_zone.h#L33)):*
> ```objc
>  * A representation of a Google Cloud zone.
>  */
> class CloudZone {
> ```

This can just be a struct



_Code quote:_
```objc
CloudZone
```
___
*[google/cloud/pubsublite/internal/cloud_zone.h, line 35 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Myx06zw42uSLARHjhBJ:-Myx06zw42uSLARHjhBK:b-k7w5s7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/internal/cloud_zone.h#L35)):*
> ```objc
> class CloudZone {
>  public:
>   explicit CloudZone(CloudRegion region, char zone_id)
> ```

remove the constructor, just use a struct



_Code quote:_
```objc
  explicit CloudZone(CloudRegion region, char zone_id)
      : region_{std::move(region)}, zone_id_{zone_id} {}
```
___
*[google/cloud/pubsublite/internal/cloud_zone.h, line 38 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Myx-f4CDp3RLTXQNWP6:-Myx-f4CDp3RLTXQNWP7:b-xink3t) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/internal/cloud_zone.h#L38)):*
> ```objc
>       : region_{std::move(region)}, zone_id_{zone_id} {}
> 
>   CloudRegion const& GetCloudRegion() const { return region_; }
> ```

remove all the "Get" prefixes



_Code quote:_
```objc
Get
```
___
*[google/cloud/pubsublite/internal/cloud_zone.h, line 51 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Myx-r3U25pcNhU1LOpL:-Myx-r3U25pcNhU1LOpM:b-gsavh6) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/internal/cloud_zone.h#L51)):*
> ```objc
>  * <location>-<direction><number>-<letter>
>  */
> StatusOr<CloudZone> ParseCloudZone(std::string const& zone) {
> ```

make this a static member function, CloudZone::Parse



_Code quote:_
```objc
ParseCloudZone
```
___
*[google/cloud/pubsublite/internal/cloud_zone.h, line 54 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Myx-wqx-2tPx_e1vlDp:-Myx-wqx-2tPx_e1vlDq:b-hvjfo9) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/internal/cloud_zone.h#L54)):*
> ```objc
>   std::vector<std::string> splits = absl::StrSplit(zone, '-');
>   if (splits.size() != 3 || splits[2].length() != 1) {
>     return StatusOr<CloudZone>{
> ```

not needed. you can just return the Status



_Code quote:_
```objc
StatusOr<CloudZone>{
```
___
*[google/cloud/pubsublite/internal/cloud_zone.h, line 57 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Myx0-1b4cDIUpq5aDxK:-Myx0-1b4cDIUpq5aDxL:b-xj9zqb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/internal/cloud_zone.h#L57)):*
> ```objc
>         Status{StatusCode::kInvalidArgument, "Invalid Zone Name"}};
>   }
>   return StatusOr<CloudZone>{CloudZone{
> ```

Not needed just return the CloudZone



_Code quote:_
```objc
StatusOr<CloudZone>{
```
___
*[google/cloud/pubsublite/internal/location.h, line 37 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Myx-MCq107olxVNZ0UE:-Myx-MCq107olxVNZ0UF:bwso6mr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/internal/location.h#L37)):*
> ```objc
>   explicit Location(CloudZone zone) : value_{std::move(zone)} {}
> 
>   bool HasCloudRegion() const {
> ```

get rid of these, and the accessors. You only need two functions: ToString, and GetCloudRegion, which should work regardless of whether it holds a region or a zone.



_Code quote:_
```objc
  bool HasCloudRegion() const {
    return absl::holds_alternative<CloudRegion>(value_);
  }

  bool HasCloudZone() const {
    return absl::holds_alternative<CloudZone>(value_);
  }
```
___
*[google/cloud/pubsublite/internal/location.h, line 69 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Myx-ZH99sMWRZEoqWve:-Myx-ZH99sMWRZEoqWvf:b-rb2npg) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/990a851a3aa9d0555d813d5dbc8b8b01cf5acdf2/google/cloud/pubsublite/internal/location.h#L69)):*
> ```objc
>     return Location{*possible_zone};
>   }
>   return Location{CloudRegion{location}};
> ```

you should probably also have a parse function for this



_Code quote:_
```objc
CloudRegion{location}
```

**Comment by coryan** [comment]
drive-by: we do not use structs like `CloudRegion` or `CloudZone` anywhere else in the library, and I am not sure we need `Location` either (maybe, but not convinced, and when in doubt we prefer to **not** introduce APIs).

I think we would all save a lot of angst if you wrote a design document describing the user-facing APIs before you start sending PRs for them.

File: google/cloud/pubsublite/topic_test.cc [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  std::string location = "location";
+  std::string topic_name = "topic_name";
+
+  Topic topic{project, location, topic_name};
+  EXPECT_EQ(project, topic.project());
+  EXPECT_EQ(location, topic.location());
+  EXPECT_EQ(topic_name, topic.topic_name());
+  EXPECT_EQ(
+      topic.FullName(),
+      absl::StrJoin(std::vector<std::string>{"projects", project, "locations",
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/topic_test.cc:40]: just type this out instead

**Review by dpcollins-google — CHANGES_REQUESTED** [review]
> Reviewed 1 of 5 files at r10, 1 of 5 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591)* status: 2 of 18 files reviewed, 5 unresolved discussions (waiting on @18suresha and @dpcollins-google)

___
*[google/cloud/pubsublite/message_metadata.h, line 46 at r12](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MzLYpYYDNoJamyp77_f:-MzLYpYYDNoJamyp77_g:bnklwbn) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0a1ce8500bfa4fb67df6dd90c6ce5e64b240375e/google/cloud/pubsublite/message_metadata.h#L46)):*
> ```objc
> 
>   /**
>    * Parses a string into a `MessageMetadata` object with the expected format
> ```

Parses a string into a `MessageMetadata` object. The formatting of this string is not stable cross-binary.



_Code quote:_
```objc
   * Parses a string into a `MessageMetadata` object with the expected format
   * "<partition>:<offset of cursor>"
```
___
*[google/cloud/pubsublite/message_metadata.h, line 64 at r12](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MzLZ-ai6P2Sm5ke1e4g:-MzLZ-ai6P2Sm5ke1e4h:b3fxu48) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0a1ce8500bfa4fb67df6dd90c6ce5e64b240375e/google/cloud/pubsublite/message_metadata.h#L64)):*
> ```objc
> 
>   /**
>    * Serializes current object as "<partition>:<offset of cursor>".
> ```

Ditto on the formatting comment above. Don't mention the specific formatting.



_Code quote:_
```objc
 Serializes current object as "<partition>:<offset of cursor>".
```
___
*[google/cloud/pubsublite/message_metadata_test.cc, line 69 at r12](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MzLZEEmB9LDj-OtDcwJ:-MzLZEEmB9LDj-OtDcwK:b-81qsuo) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0a1ce8500bfa4fb67df6dd90c6ce5e64b240375e/google/cloud/pubsublite/message_metadata_test.cc#L69)):*
> ```cpp
> }
> 
> TEST(MessageMetadata, Serialize) {
> ```

Add a round trip test.

```
MessageMetadata mm{...};
EXPECT_EQ(MessageMetadata::Parse(mm.Serialize()).value(), mm);
```



_Code quote:_
```cpp
Serialize
```
___
*[google/cloud/pubsublite/internal/cloud_region.h, line 33 at r12](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MzLZiyx1NezBz6e3h32:-MzLZiyy00gZvA3Qorbf:b-cedrqy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0a1ce8500bfa4fb67df6dd90c6ce5e64b240375e/google/cloud/pubsublite/internal/cloud_region.h#L33)):*
> ```objc
>  */
> struct CloudRegion {
>   explicit CloudRegion(std::string region) : region{std::move(region)} {}
> ```

Get rid of the constructors for the Cloud{Region,Zone,Location} types, you can just use the struct constructor that exists by default with curly braces



_Code quote:_
```objc
explicit CloudRegion(std::string region) : region{std::move(region)} {}
```
___
*[google/cloud/pubsublite/internal/location.h, line 54 at r12](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MzL_7eJ6sE0FPATW5fR:-MzL_7eJ6sE0FPATW5fS:b-rzw97a) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0a1ce8500bfa4fb67df6dd90c6ce5e64b240375e/google/cloud/pubsublite/internal/location.h#L54)):*
> ```objc
>     CloudZone cloud_zone = absl::get<CloudZone>(value_);
>     return absl::StrCat(cloud_zone.region.region, "-",
>                         std::string(1, cloud_zone.zone_id));
> ```

nit, add ToString methods to both of these classes?



_Code quote:_
```objc
    if (absl::holds_alternative<CloudRegion>(value_)) {
      return absl::get<CloudRegion>(value_).region;
    }
    CloudZone cloud_zone = absl::get<CloudZone>(value_);
    return absl::StrCat(cloud_zone.region.region, "-",
                        std::string(1, cloud_zone.zone_id));
```

**Review by dpcollins-google — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591)* status: 2 of 18 files reviewed, 3 unresolved discussions (waiting on @18suresha and @dpcollins-google)

___
*[google/cloud/pubsublite/message_metadata_test.cc, line 69 at r12](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MzLZEEmB9LDj-OtDcwJ:-MzLhYf5CIRhEkYmwUNr:bukc70i) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0a1ce8500bfa4fb67df6dd90c6ce5e64b240375e/google/cloud/pubsublite/message_metadata_test.cc#L69)):*
<details><summary><i>Previously, dpcollins-google wrote…</i></summary><blockquote>

Add a round trip test.

```
MessageMetadata mm{...};
EXPECT_EQ(MessageMetadata::Parse(mm.Serialize()).value(), mm);
```
</blockquote></details>

The round trip test doesn't test the round trip I wanted you to test: that the serialized result when deserialized is the same. Please fix?
___
*[google/cloud/pubsublite/internal/cloud_region.h, line 43 at r13](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MzLhk2aA1rLFpd48NFF:-MzLhk2aA1rLFpd48NFG:b-3bs0ks) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d391b3df9b4726998efe020d8e5f11f94448bdba/google/cloud/pubsublite/internal/cloud_region.h#L43)):*
> ```objc
>   std::string ToString() const { return region; }
> 
>   std::string const region;
> ```

not const on struct members, you can have a const CloudRegion if you want one



_Code quote:_
```objc
const
```
___
*[google/cloud/pubsublite/internal/cloud_zone.h, line 52 at r13](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MzLhpyv9I9CdGX9xZoq:-MzLhpyv9I9CdGX9xZor:bfiu1kk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d391b3df9b4726998efe020d8e5f11f94448bdba/google/cloud/pubsublite/internal/cloud_zone.h#L52)):*
> ```objc
> 
>   CloudRegion const region;
>   char const zone_id;
> ```

ditto all of these
___
*[google/cloud/pubsublite/internal/location.h, line 54 at r12](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MzL_7eJ6sE0FPATW5fR:-MzLiBdfCPq4M5C59IzG:b31po3z) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0a1ce8500bfa4fb67df6dd90c6ce5e64b240375e/google/cloud/pubsublite/internal/location.h#L54)):*
<details><summary><i>Previously, dpcollins-google wrote…</i></summary><blockquote>

nit, add ToString methods to both of these classes?
</blockquote></details>

The expectation was you remove the formatting from here as well, can we do that?

**Review by 18suresha — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591)* status: 2 of 18 files reviewed, 3 unresolved discussions (waiting on @18suresha and @dpcollins-google)

___
*[google/cloud/pubsublite/message_metadata_test.cc, line 69 at r12](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-MzLZEEmB9LDj-OtDcwJ:-MzM956F5d4LV-0RdXPY:b-6unan3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0a1ce8500bfa4fb67df6dd90c6ce5e64b240375e/google/cloud/pubsublite/message_metadata_test.cc#L69)):*
<details><summary><i>Previously, dpcollins-google wrote…</i></summary><blockquote>

The round trip test doesn't test the round trip I wanted you to test: that the serialized result when deserialized is the same. Please fix?
</blockquote></details>

oh sorry I didn't see the code you pasted

File: google/cloud/pubsublite/internal/location.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+/**
+ * A wrapper around a Google Cloud Location which can be either a `CloudRegion`
+ * or `CloudZone`.
+ */
+class Location {
```
Comment (coryan) [inline google/cloud/pubsublite/internal/location.h:35]: This is not in your design doc, and I suspect we won't need it.  Please describe it there and/or remove it from this PR until it is discussed in the design doc.
Ditto for `CloudRegion` and `CloudZone`.

Comment (18suresha) [reply google/cloud/pubsublite/internal/location.h:35]: These aren't user-facing though? We need these data representations in order to parse region names from the user-provided `Topic` with which we'll create publisher connections.

**Comment by coryan** [comment]
___
*[google/cloud/pubsublite/internal/location.h, line 35 at r17](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Mywa-ImJzyKdZc2HqC_-r17-35:-MzMf497CZTtm0jN7CeS:bfhdiq9) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/792dd6144895f68093e3787c186e419e7cc97f07/google/cloud/pubsublite/internal/location.h#L35)):*
<details><summary><i>Previously, 18suresha (Adarsh Suresh) wrote…</i></summary><blockquote>

These aren't user-facing though? We need these data representations in order to parse region names from the user-provided `Topic` with which we'll create publisher connections.
</blockquote></details>

Doh!  I had done a cursory read of this PR and did not notice you moved them.  With that said, they probably no longer belong in this PR (given its description)?

**Review by 18suresha — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591)* status: 2 of 18 files reviewed, 4 unresolved discussions (waiting on @18suresha, @coryan, and @dpcollins-google)

___
*[google/cloud/pubsublite/internal/location.h, line 35 at r17](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591#-Mywa-ImJzyKdZc2HqC_-r17-35:-MzMgyAh5O9WYNoLaTaP:b4kxppg) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/792dd6144895f68093e3787c186e419e7cc97f07/google/cloud/pubsublite/internal/location.h#L35)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Doh!  I had done a cursory read of this PR and did not notice you moved them.  With that said, they probably no longer belong in this PR (given its description)?
</blockquote></details>

Oh yea there is a discrepancy with the description (just updated it)
I'd prefer to keep it all in this PR since all the changes are still basic data representations and parsing

File: google/cloud/pubsublite/topic_test.cc [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+namespace pubsublite_internal {
+namespace {
+
+TEST(Topic, BasicTopic) {
+  std::string project = "project";
+  std::string location = "location";
+  std::string topic_name = "topic_name";
+
+  Topic topic{project, location, topic_name};
+  EXPECT_EQ(project, topic.project());
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/topic_test.cc:35]: nit. flip these expectations so they're not yoda conditions

**Comment by dpcollins-google** [comment]
> If that, in other libraries we do not define wrappers for things that are just strings. I am not sure I see the need for them here.

In this case, CloudRegion is needed for parity with CloudZone. You need to put <something> in the variant in the other slot, and having it be a proper type as opposed to just std::string makes it easier to read.

**Review by coryan — APPROVED** [review]
> Reviewed 4 of 6 files at r20, 7 of 9 files at r21, 2 of 2 files at r22, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8591)* status: all files reviewed, 4 unresolved discussions (waiting on @18suresha)

**Pull request merged**